File: javascript/src/aplicacao/AgendamentoServiceImpl.js
import { AgendaService } from '../porta/AgendaService.js';
import { Consulta } from '../dominio/Consulta.js';
import { Mensagem } from '../dominio/Mensagem.js';
import { RegistroUsuarios } from './RegistroUsuarios.js';

/**
 * DIP puro: os tres colaboradores chegam pelo construtor como CONTRATO.
 * Trocar memoria por Postgres, ou email por whatsapp, nao mexe numa linha daqui.
 */
export class AgendamentoServiceImpl extends AgendaService {
  #repo;
  #notificador;
  #politica;

  constructor(repo, notificador, politica) {
    super();
    this.#repo = repo;
    this.#notificador = notificador;
    this.#politica = politica;
  }

  agendar(paciente, medico, dataHora, modalidade) {
    if (!this.#politica.estaLivre(medico, dataHora)) {
      throw new Error(`Medico indisponivel em ${dataHora.toISOString()}`);
    }
    const consulta = new Consulta(paciente, medico, dataHora, modalidade);
    consulta.confirmar();
    this.#repo.salvar(consulta);
    this.#avisar(paciente, 'Consulta confirmada',
      `Sua consulta foi marcada para ${dataHora.toLocaleString('pt-BR')}.`);
    return consulta;
  }

  cancelar(id, motivo) {
    const consulta = this.#repo.buscarPorId(id);
    consulta.cancelar(motivo);
    this.#repo.salvar(consulta);
    this.#avisar(consulta.paciente, 'Consulta cancelada', `Motivo: ${motivo}`);
  }

  remarcar(id, nova) {
    const consulta = this.#repo.buscarPorId(id);
    if (!this.#politica.estaLivre(consulta.medico, nova)) {
      throw new Error(`Medico indisponivel em ${nova.toISOString()}`);
    }
    consulta.remarcar(nova);
    this.#repo.salvar(consulta);
    this.#avisar(consulta.paciente, 'Consulta remarcada',
      `Novo horario: ${nova.toLocaleString('pt-BR')}`);
  }

  #avisar(paciente, titulo, corpo) {
    const dono = RegistroUsuarios.donoDoPerfil(paciente);
    if (dono) {
      this.#notificador.notificar(dono, new Mensagem(titulo, corpo));
    }
  }
}

File: java/src/br/unifor/facilitador/aplicacao/AgendamentoServiceImpl.java
package br.unifor.facilitador.aplicacao;

import br.unifor.facilitador.dominio.*;
import br.unifor.facilitador.porta.*;
import java.time.LocalDateTime;
import java.util.UUID;

/**
 * DIP puro: os tres colaboradores chegam pelo construtor como INTERFACE.
 * Trocar Postgres por memoria, ou email por whatsapp, nao mexe numa linha daqui.
 */
public class AgendamentoServiceImpl implements AgendaService {

    private final ConsultaRepository repo;
    private final Notificador notificador;
    private final PoliticaDisponibilidade politica;

    public AgendamentoServiceImpl(ConsultaRepository repo, Notificador notificador,
                                  PoliticaDisponibilidade politica) {
        this.repo = repo;
        this.notificador = notificador;
        this.politica = politica;
    }

    @Override
    public Consulta agendar(Paciente p, Medico m, LocalDateTime dt, Modalidade modalidade) {
        if (!politica.estaLivre(m, dt)) {
            throw new IllegalArgumentException("Medico indisponivel em " + dt);
        }
        Consulta c = new Consulta(p, m, dt, modalidade);
        c.confirmar();
        repo.salvar(c);
        avisar(p, "Consulta confirmada", "Sua consulta foi marcada para " + dt + ".");
        return c;
    }

    @Override
    public void cancelar(UUID id, String motivo) {
        Consulta c = repo.buscarPorId(id);
        c.cancelar(motivo);
        repo.salvar(c);
        avisar(c.getPaciente(), "Consulta cancelada", "Motivo: " + motivo);
    }

    @Override
    public void remarcar(UUID id, LocalDateTime nova) {
        Consulta c = repo.buscarPorId(id);
        if (!politica.estaLivre(c.getMedico(), nova)) {
            throw new IllegalArgumentException("Medico indisponivel em " + nova);
        }
        c.remarcar(nova);
        repo.salvar(c);
        avisar(c.getPaciente(), "Consulta remarcada", "Novo horario: " + nova);
    }

    private void avisar(Paciente p, String titulo, String corpo) {
        Usuario dono = RegistroUsuarios.donoDoPerfil(p);
        if (dono != null) {
            notificador.notificar(dono, new Mensagem(titulo, corpo));
        }
    }
}
